torch/distributed/_tensor/experimental/attention.py

```diff
@@ -246,37 +322,61 @@ def _templated_ring_attention(
     key = key.contiguous()
     value = value.contiguous()
 
-    sdpa_merger = _SDPAMerger(_convert_to_f32)
+    sdpa_merger = _SDPAMerger(_cp_options.convert_to_f32, seq_dim=seq_dim)
 
     rest: List[Any]
     out: torch.Tensor
     logsumexp: torch.Tensor
 
     for i in range(size):
-        # overlap communication with compute
         if next_kv is not None:
+            # Wait for the kv from the (cp_rank - 1) rank.
             next_kv = _maybe_wait(next_kv)
             key = next_kv[: key.numel()].reshape(key.shape)
             value = next_kv[key.numel() :].reshape(value.shape)
 
         if i < (size - 1):
+            # Send the k, v to the next rank
             next_kv = torch.cat([key.flatten(), value.flatten()])
             next_kv = _ring_rotate(next_kv, pg, send_to_next=True)
 
         is_causal_behavior = _is_causal_behavior(
             rank=rank, world_size=size, i=i, is_causal=is_causal
         )
 
-        if is_causal_behavior != _CausalBehavior.SKIP:
-            out, logsumexp, *rest = op(
+        if is_causal_behavior == _CausalBehavior.SKIP:
+            continue
+
+        if i == 0 or (not _cp_options.enable_load_balance or not is_causal):
+            # We need to do SDPA with the full local q, k, v.
```

Comment:
we discussed offline about adding a more detailed 'picture' of the round-robin splitting of QKV and using that to help explain the if/elif/else.

But one minor change  that would help, "full local" can be replaced with "both local chunks of" to avoid ambiguity